app/services/onlinetesting/OnlineTestService.scala

```diff
@@ -323,6 +325,52 @@ trait OnlineTestService {
     }
   }
 
+  def markAsStarted(cubiksUserId: Int): Future[Unit] = {
+    updateTestPhase1(cubiksUserId, t => t.copy(startedDateTime = Some(DateTimeFactory.nowLocalTimeZone)))
+  }
+
+  def markAsCompleted(cubiksUserId: Int): Future[Unit] = {
+    updateTestPhase1(cubiksUserId, t => t.copy(completedDateTime = Some(DateTimeFactory.nowLocalTimeZone)))
+  }
+
+  def markAsCompleted(token: String): Future[Unit] = {
+    otRepository.getPhase1TestProfileByToken(token).flatMap { p =>
+      val test = p.tests.find(_.token == token).get
+      markAsCompleted(test.cubiksUserId)
+    }
+  }
+
+  def markAsReportReadyToDownload(cubiksUserId: Int, reportReady: Phase1TestResultReady): Future[Unit] = {
+    updateTestPhase1(cubiksUserId,
+      t => t.copy(
+        resultsReadyToDownload = reportReady.reportStatus == "Ready",
+        reportId = reportReady.reportId,
+        reportLinkURL = reportReady.reportLinkURL,
+        reportStatus = Some(reportReady.reportStatus)
+      )
+    )
+  }
+
+  private def updateTestPhase1(cubiksUserId: Int, update: Phase1Test => Phase1Test): Future[Unit] = {
+    def createUpdateTestGroup(p: Phase1TestProfileWithAppId): (String, Phase1TestProfile) = {
+      val testGroup = p.phase1TestProfile
+      require(testGroup.tests.count(_.cubiksUserId == cubiksUserId) == 1)
+      val appId = p.applicationId
+      val updatedTestsWithTodaysDate = testGroup.tests.collect {
+        case t if t.cubiksUserId == cubiksUserId => update(t)
+        case t => t
+      }
+      val updatedTestGroup = testGroup.copy(tests = updatedTestsWithTodaysDate)
+      (appId, updatedTestGroup)
+    }
+
+    for {
+      p <- otRepository.getPhase1TestProfileByCubiksId(cubiksUserId)
+      (appId, updatedTestGroup) = createUpdateTestGroup(p)
```

Comment:
There is a potential race condition here. Did we look at the possibility of using findAndModify rather than pulling the record back to the service?